drivers/disk/ffatdisk.c

```diff
@@ -0,0 +1,976 @@
+		LOG_DBG("Read FSI (%u)", sector);
+
+		return;
+	}
+
+	if (IN_RANGE(sector, config->fat1_start, config->fat2_start - 1)) {
+		fat_sector = sector - config->fat1_start;
+		LOG_DBG("Read FAT1 sector %u", fat_sector);
+		config->ffat_read(dev, buf, fat_sector);
+
+		return;
+	}
+
+	if (IN_RANGE(sector, config->fat2_start, config->rd_start - 1)) {
+		fat_sector = sector - config->fat2_start;
+		LOG_DBG("Read FAT2 sector %u", fat_sector);
+		config->ffat_read(dev, buf, fat_sector);
+
+		return;
+	}
+
+	if (IN_RANGE(sector, config->rd_start, config->data_start - 1)) {
+		ffat_read_rd(dev, buf, sector);
+
+		return;
+	}
+
+	if (IN_RANGE(sector, config->data_start, config->sector_count - 1)) {
+		ffat_read_file(dev, buf, sector);
+
+		return;
+	}
+}
+
+static int ffatdisk_access_read(struct disk_info *const disk, uint8_t *const buf,
+				const uint32_t sector, const uint32_t count)
+{
+	const struct device *dev = disk->dev;
+	const struct ffatdisk_config *config = dev->config;
+	uint32_t sector_max = sector + count;
+
+	if (sector_max < sector || sector_max > config->sector_count) {
+		LOG_ERR("Sector %u is outside the range %u",
+			sector_max, config->sector_count);
+		return -EIO;
+	}
+
+	for (uint32_t i = sector; i < sector_max; i++) {
+		ffat_read_sector(disk, buf, i);
+	}
+
+	return 0;
+}
+
+static int ffat_write_file(const struct device *dev,
+			   const uint8_t *const buf, const uint32_t sector)
+{
+	struct ffatdisk_data *const data = dev->data;
+	uint32_t byts_per_sec = bs_get_byts_per_sec(dev);
+	struct ffat_file *f;
+	uint32_t f_block = 0UL;
+
+	if (data->clusters_free) {
+		/*
+		 * If there are free clusters on the volume,
+		 * the filesystem driver can write to any of the free
+		 * clusters and we cannot determine the sector number
+		 * of the file. Therefore, there should be only one file
+		 * with write callback. To get the exact sector (or block)
+		 * number and other metadata, backends need to encapsulate
+		 * the payload, which we do not care about at all.
+		 */
+		f = ffat_get_file_any_wr_cb(dev);
+	} else {
+		/*
+		 * If there are no free clusters on the volume,
+		 * we can determine the sector index of the file.
+		 * This is nice, but less practical because some OS
+		 * do not want to overwrite the file if there is no space.
+		 */
+		f = ffat_get_file_and_block(dev, sector, &f_block);
+	}
+
+	if (f != NULL && f->wr_cb != NULL) {
+		f->wr_cb(f, f_block, buf, byts_per_sec);
+		LOG_DBG("Write file block %u (%u)", f_block, sector);
+	}
+
+	return 0;
+}
+
+static int ffat_write_sector(struct disk_info *const disk,
+			     const uint8_t *const buf, const uint32_t sector)
+{
+	const struct device *dev = disk->dev;
+	const struct ffatdisk_config *config = dev->config;
+	uint32_t fat_sector;
+
+	/*
+	 * For now, we ignore write accesses from the (host) filesystem driver
+	 * to all sectors except the data area. Perhaps we can use some of these
+	 * to implement a mounted/unmounted state indication callback.
+	 */
```

Comment:
How this is not leading to absolutely bizzare and pretty much impossible to debug issues?

The idea of abusing FAT file system for specific uses that can work is https://github.com/microsoft/uf2. Note that UF2 operates on different principle and is pretty much limited to write only.

The fakefat in my opinion cannot really work reliably, because host pretty much has complete control over MSC disks.

I believe MTP is the solution for exporting fixed amount of files, not MSC.

Reply:
> How this is not leading to absolutely bizzare and pretty much impossible to debug issues?

What issues and for what use cases? The write accesses can be ignored because we do not change e.g. the FAT, what is emulated anyway.

> The idea of abusing FAT file system for specific uses that can work is https://github.com/microsoft/uf2. Note that UF2 operates on different principle and is pretty much limited to write only.

This PR uses a similar principle (also similar to virtual_fs in daplink). One of my intentions is to use it to provide similar functionality as the DAPLink firmware. And I am pretty sure you can read emulated files in UF2 (and daplink's virtual_fs).

> 
> The fakefat in my opinion cannot really work reliably, because host pretty much has complete control over MSC disks.
> 
> I believe MTP is the solution for exporting fixed amount of files, not MSC.

It works very reliably for what it was developed for. It is not there to export arbitrary file systems or to be something like MTP.

Reply:
Virtual fs in daplink is actively checking for problems (e.g. data being sent sequential) while this implementation does not.

UF2 is operating on different principle. UF2 only assumes that files are written on sector boundary (i.e. file does not start in the middle of sector) and has plenty of padding.